model/textparse/nhcbparse.go

```diff
@@ -0,0 +1,333 @@
+	p.lastHistogramLabelsHash, _ = p.lset.HashWithoutLabels(p.hBuffer, labels.BucketLabel)
+}
+
+// handleClassicHistogramSeries collates the classic histogram series to be converted to NHCB
+// if it is actually a classic histogram series (and not a normal float series) and if there
+// isn't already a native histogram with the same name (assuming it is always processed
+// right before the classic histograms) and returns true if the collation was done.
+func (p *NHCBParser) handleClassicHistogramSeries(lset labels.Labels) bool {
+	if p.typ != model.MetricTypeHistogram {
+		return false
+	}
+	mName := lset.Get(labels.MetricName)
+	// Sanity check to ensure that the TYPE metadata entry name is the same as the base name.
+	if convertnhcb.GetHistogramMetricBaseName(mName) != string(p.bName) {
+		return false
+	}
+	switch {
+	case strings.HasSuffix(mName, "_bucket") && lset.Has(labels.BucketLabel):
+		le, err := strconv.ParseFloat(lset.Get(labels.BucketLabel), 64)
+		if err == nil && !math.IsNaN(le) {
+			p.processClassicHistogramSeries(lset, "_bucket", func(hist *convertnhcb.TempHistogram) {
+				hist.BucketCounts[le] = p.value
+			})
+			return true
+		}
+	case strings.HasSuffix(mName, "_count"):
+		p.processClassicHistogramSeries(lset, "_count", func(hist *convertnhcb.TempHistogram) {
+			hist.Count = p.value
+		})
+		return true
+	case strings.HasSuffix(mName, "_sum"):
+		p.processClassicHistogramSeries(lset, "_sum", func(hist *convertnhcb.TempHistogram) {
+			hist.Sum = p.value
+		})
+		return true
+	}
+	return false
+}
+
+func (p *NHCBParser) processClassicHistogramSeries(lset labels.Labels, suffix string, updateHist func(*convertnhcb.TempHistogram)) {
+	p.isCollationInProgress = true
+	p.tempLsetNHCB = convertnhcb.GetHistogramMetricBase(lset, suffix)
+	p.storeExemplars()
+	updateHist(&p.tempNHCB)
+}
+
+func (p *NHCBParser) storeExemplars() {
+	for ex := p.nextExemplarPtr(); p.parser.Exemplar(ex); ex = p.nextExemplarPtr() {
+		p.tempExemplarCount++
+	}
+}
+
+func (p *NHCBParser) nextExemplarPtr() *exemplar.Exemplar {
+	switch {
+	case p.tempExemplarCount == len(p.tempExemplars)-1:
+		// Reuse the previously allocated exemplar, it was not filled up.
+	case len(p.tempExemplars) == cap(p.tempExemplars):
+		// Let the runtime grow the slice.
+		p.tempExemplars = append(p.tempExemplars, exemplar.Exemplar{})
+	default:
+		// Take the next element into use.
+		p.tempExemplars = p.tempExemplars[:len(p.tempExemplars)+1]
+	}
+	return &p.tempExemplars[len(p.tempExemplars)-1]
+}
+
+func (p *NHCBParser) swapExemplars() {
+	p.exemplars = p.tempExemplars[:p.tempExemplarCount]
+	p.tempExemplars = p.tempExemplars[:0]
+	p.tempExemplarCount = 0
+}
+
+// processNHCB converts the collated classic histogram series to NHCB and caches the info
+// to be returned to callers.
+func (p *NHCBParser) processNHCB() bool {
+	if !p.isCollationInProgress {
+		return false
+	}
+	ub := make([]float64, 0, len(p.tempNHCB.BucketCounts))
```

Comment:
some slice to potentially reuse? (feel free to ignore if not visible in any profile)

Reply:
I'm prioritizing correctness for now.